platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/AbstractCollectionProperty.java

```diff
@@ -486,4 +497,61 @@ public ValueProducer getProducer() {
             return left.getProducer().plus(right.getProducer());
         }
     }
+
+    protected CollectionSupplierGuard<T, C> guardSupplier(CollectionSupplier<T, C> supplier) {
+        return new CollectionSupplierGuard<>(this, supplier);
+    }
+
+    protected static final class CollectionSupplierGuard<T, C extends Collection<T>> implements CollectionSupplier<T, C>, GuardedData<CollectionSupplier<T, C>> {
+        private final ProviderInternal<?> owner;
+        private final CollectionSupplier<T, C> supplier;
+
+        public CollectionSupplierGuard(ProviderInternal<?> owner, CollectionSupplier<T, C> supplier) {
+            this.owner = owner;
+            this.supplier = supplier;
+        }
+
+        @Override
+        public ProviderInternal<?> getOwner() {
+            return owner;
+        }
+
+        @Override
+        public CollectionSupplier<T, C> unsafeGet() {
+            return supplier;
+        }
+
+        @Override
+        public Value<? extends C> calculateValue(ValueConsumer consumer) {
+            try (EvaluationContext.ScopeContext ignore = EvaluationContext.current().enter(owner)) {
+                return supplier.calculateValue(consumer);
+            }
+        }
+
+        @Override
+        public CollectionSupplierGuard<T, C> plus(Collector<T> collector) {
+            return new CollectionSupplierGuard<>(owner, supplier.plus(collector));
+        }
+
+        @Override
+        public ExecutionTimeValue<? extends C> calculateExecutionTimeValue() {
```

Comment:
❓ 

This is feedback that should have gone into the spec. Sorry if it is too late. I like the design and implementation of the subsystem for cycle detection, but hard to ignore the size of the design footprint required to prepare evaluation participants to be protected against cycles.  

Please forgive me if I am being overly naive (I often am), but did you consider a design with a smaller footprint, that did not involve decorators?